hydroflow_lang/src/graph/serde_graph.rs

```diff
@@ -0,0 +1,174 @@
+            }
+        }
+        Ok(())
+    }
+
+    pub fn serde_to_dot(&self) -> String {
+        let mut output = String::new();
+        self.serde_write_dot(&mut output).unwrap();
+        output
+    }
+
+    pub fn serde_write_dot_edge(
+        &self,
+        src: GraphNodeId,
+        dst: GraphNodeId,
+        tab: usize,
+        w: &mut impl std::fmt::Write,
+    ) -> std::fmt::Result {
+        writeln!(
+            w,
+            "{:t$}{} -> {}",
+            "",
+            src.data().as_ffi(),
+            dst.data().as_ffi(),
+            t = tab,
+        )?;
+        Ok(())
+    }
+    pub fn serde_write_dot(&self, w: &mut impl std::fmt::Write) -> std::fmt::Result {
+        writeln!(w, "digraph {{")?;
+        let mut tab: usize = 2;
+
+        // write out nodes
+        writeln!(w, "{:t$}{{", "", t = tab)?;
+        tab += 2;
+        writeln!(w, "{:t$}node [shape=box]", "", t = tab)?;
+        for (node_id, text) in self.nodes.iter() {
+            let nm = text.replace('"', "\\\"").replace('\n', "\\l");
+            let label = format!("{}", node_id.data().as_ffi());
+            write!(
+                w,
+                "{:t$}{} [label=\"{}{}\"",
+                "",
+                label,
+                nm.clone(),
+                match nm.contains("\\l") {
+                    // if contains linebreak left-justify by appending another "\\l"
+                    true => "\\l",
+                    false => "",
+                },
+                t = tab
+            )?;
+            if self.handoffs.contains_key(node_id) {
+                write!(w, ", shape=diamond")?;
+            }
+            writeln!(w, "]")?;
+        }
+        tab -= 2;
+        writeln!(w, "{:t$}}}", "", t = tab)?;
+
+        // write out edges per subgraph
+        for (sg_id, nodes) in &self.subgraph_nodes {
+            // let strt = es.stratum;
+            writeln!(
+                w,
+                "{:t$}subgraph \"cluster stratum NUMBER UNKNOWN\" {{",
+                "",
+                t = tab,
+            )?;
+            tab += 2;
+            writeln!(w, "{:t$}label = \"Stratum NUMBER UNKNOWN\"", "", t = tab,)?;
+            writeln!(
+                w,
+                "{:t$}subgraph \"cluster {}\" {{",
+                "",
+                sg_id.data().as_ffi(),
+                t = tab
+            )?;
+            tab += 2;
+            writeln!(
+                w,
+                "{:t$}label = \"sg_{}\"",
+                "",
+                sg_id.data().as_ffi(),
+                t = tab
+            )?;
+            let empty = vec![];
+            for src in nodes {
+                let dests = self.edges.get(*src).unwrap_or(&empty);
+                for dst in dests {
+                    if !self.handoffs.contains_key(*src) && !self.handoffs.contains_key(*dst) {
+                        self.serde_write_dot_edge(*src, *dst, tab, w)?;
+                    }
+                }
+            }
```

Comment:
`for &src in nodes` and `for &dst in dests` will let you remove all the dereference `*`s